spec/src/main/resources/META-INF/smithy/bsp/extensions/cancel.smithy

```diff
@@ -0,0 +1,33 @@
+$version: "2"
+
+namespace bsp.cancel
+
+use traits#jsonNotification
+use traits#jsonRPC
+use traits#untaggedUnion
+
+@jsonRPC
+service CancelExtension {
```

Comment:
Why is this an extension and not part of the core protocol?

Reply:
@agluszak proposed this solution to prevent overriding lsp `Cancel` implementation and it seems reasonable

> I think however that due to how cancellation is implemented in lsp4j this should be extracted to a separate interface (otherwise the default cancellation mechanism would be overridden). This can be done either in the lsp4j code generator or in the smithy definition (this would have to live in a separate "extension" then)

Reply:
So this is a copy of the LSP cancel protocol.

Do we still depend on the lsp4j library? If so, do we have a plan to remove it?

Reply:
> Why is this an extension and not part of the core protocol?

Because some libraries don't have to implement it explicitly.

> So this is a copy of the LSP cancel protocol.

Yes

> Do we still depend on the lsp4j library?

Partially. The lsp4j library consists of two separate things. One is the JsonRpc library used by bsp4j. Cancellation handling is defined inside lsp4j.jsonrpc. The other thing is the support library for LSP (in the same sense as bsp4j is a Java support library for BSP). We never used it.

> If so, do we have a plan to remove it?

I don't think so.

Reply:
Ok so we need to implement cancellation in some languages but not in Java because we have lsp4j.

Reply:
Exactly